lib/spack/spack/package.py

```diff
@@ -467,7 +467,19 @@ def test_log_pathname(test_stage, spec):
                         'test-{0}-out.txt'.format(TestSuite.test_pkg_id(spec)))
 
 
-class PackageBase(six.with_metaclass(PackageMeta, PackageViewMixin, object)):
+class RedistributionMixin(object):
+    #: Whether it should be possible to add the source of this package to a
+    #: Spack mirror. This is only false when licensing for the package
+    #: implies that public mirrors should not redistribute its source code.
+    redistribute_source = True
+
+    #: Whether it should be possible to create a binary out of an installed
+    #: instance of this package.
+    redistribute_binary = True
```

Comment:
Can we make these two attributes into a directive, e.g.:

```python
class MyPackage(Package):
    redistribute(source=True, binary=False)
```

Where the defaults are (for now) `True` and `True` but eventually (after this PR) taken from license metadata?

I could see more conditions being added here and it's an easy way to allow this to evolve without adding a lot of attributes.